### ISSUE TITLE
Can peft be used with HuggingFace's Trainer and DeepSpeed?

### ISSUE THREAD

**Issue by nanyyyyyy** [huggingface/peft#316]
I got this error when using deepspeed zero2. I am using huggingface's trainer.
```
return self.flatten(align_dense_tensors(tensor_list, alignment))
RuntimeError: torch.cat(): expected a non-empty list of Tensors
```

**Comment by wxjiao** [comment]
@nanyyyyyy Yes, we have an implementation based on the example script `run_clm.py`. Refer to https://github.com/wxjiao/ParroT for more details. BTW, we used zero3.

**Comment by Taoooo9** [comment]
> 
Hi, have you solved this problem?

**Comment by nanyyyyyy** [comment]
> Hi, have you solved this problem?

Yes. @wxjiao their implementation should work. Thanks @wxjiao btw.

**Comment by wxjiao** [comment]
@nanyyyyyy Thanks for mentioning. We just conducted a few optimzations for the previous implementation. Generally, both ZeRO2 and ZeRO3 are supported in our implementation. However, for medium-to-small LMs (e.g., 7B), I would recommend to use ZeRO2+Offload which runs faster than ZeRO3. 
For LoRA, ZeRO3 may be problematic when facing large adapers. @Taoooo9

**Comment by arash2060** [comment]
I don't follow. What has been implemented differently in https://github.com/wxjiao/ParroT? Is this a new version of transformers? The config is different? or just the script?
